File: backend/src/models/Group.js
import mongoose from 'mongoose';

const memberSchema = new mongoose.Schema({
  id: {
    type: String,
    required: true,
  },
  name: {
    type: String,
    required: true,
  },
});

const expenseSchema = new mongoose.Schema({
  id: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: true,
  },
  amount: {
    type: Number,
    required: true,
    min: 0,
  },
  paidBy: {
    type: String,
    required: true,
  },
  participants: [{
    type: String,
    required: true,
  }],
  date: {
    type: Date,
    default: Date.now,
  },
});

const groupSchema = new mongoose.Schema({
  id: {
    type: String,
    required: true,
    unique: true,
    index: true,
  },
  name: {
    type: String,
    required: true,
  },
  members: [memberSchema],
  expenses: [expenseSchema],
  userId: {
    type: String,
    required: true,
    index: true,
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
});

// Update timestamp on save
groupSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

// Index for efficient queries
groupSchema.index({ userId: 1, createdAt: -1 });

export const Group = mongoose.model('Group', groupSchema);

File: backend/models/group.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Member represents a member in a group
type Member struct {
	ID   string `bson:"id" json:"id"`
	Name string `bson:"name" json:"name"`
}

// Expense represents an expense in a group
type Expense struct {
	ID           string    `bson:"id" json:"id"`
	Description  string    `bson:"description" json:"description"`
	Amount       float64   `bson:"amount" json:"amount"`
	PaidBy       string    `bson:"paidBy" json:"paidBy"`
	Participants []string  `bson:"participants" json:"participants"`
	Date         time.Time `bson:"date" json:"date"`
}

// Group represents a group with members and expenses
type Group struct {
	ID        primitive.ObjectID `bson:"_id,omitempty" json:"_id,omitempty"`
	GroupID   string             `bson:"id" json:"id"`
	Name      string             `bson:"name" json:"name"`
	Members   []Member           `bson:"members" json:"members"`
	Expenses  []Expense          `bson:"expenses" json:"expenses"`
	UserID    string             `bson:"userId" json:"userId"`
	CreatedAt time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time          `bson:"updatedAt" json:"updatedAt"`
}

// GroupResponse is the response structure for group data
type GroupResponse struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Members   []Member  `json:"members"`
	Expenses  []Expense `json:"expenses"`
	CreatedAt time.Time `json:"createdAt"`
}
